app/os_android.go

```diff
@@ -951,3 +959,59 @@ func Java_org_gioui_Gio_scheduleMainFuncs(env *C.JNIEnv, cls C.jclass) {
 }
 
 func (_ ViewEvent) ImplementsEvent() {}
+
+var foregroundService struct {
+	once   sync.Once
+	intent C.jobject
+	mu     sync.Mutex
+	count  int
+}
+
+// StartForeground starts the foreground service
+func StartForeground(title, text string) (err error) {
+	foregroundService.mu.Lock()
+	defer foregroundService.mu.Unlock()
+	foregroundService.once.Do(func() {
+		runInJVM(javaVM(), func(env *C.JNIEnv) {
+			startForegroundService := getStaticMethodID(env, android.gioCls,
+				"startForegroundService",
+				"(Landroid/content/Context;Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;",
+			)
+			foregroundService.intent, err = callStaticObjectMethod(env, android.gioCls,
+				startForegroundService,
+				jvalue(android.appCtx),
+				jvalue(javaString(env, title)),
+				jvalue(javaString(env, text)),
+			)
+			if err == nil {
+				// get a reference across JNI sessions to the returned intent
+				foregroundService.intent = C.jni_NewGlobalRef(env, foregroundService.intent)
+			}
+		})
+	})
+	if err == nil {
+		foregroundService.count++
+	}
+	return
+}
+
+// StopForeground stops the foreground service
+func StopForeground() (err error) {
```

Comment:
A bare StopForeground makes it easy to stop a different foreground intent than intended. I suggest getting rid of foreground.Stop and change Start to

func Start(title, text string) (stop <-chan struct{}, err error)

where `close(stop)` will stop the foreground notification.

To implement this, I suggest starting a goroutine in foreground.Start that waits for stop to close. I believe you will need to replace foregroundService.count with a map to keep track of the active foregrounds.

Reply:
To match API such as `context`, it's probably better to return a closure:

func Start(title, text string) (stop func(), err error)

instead.

Reply:
I don't understand why you think there are multiple foreground processes - GioForegroundService is a singleton and just needs to be running in order for the application to not be killed when the application no longer has foreground focus. The caller is responsible for deciding when the application needs GioForegroundService running, so even if the API should permit the caller to be lazy and call StartForeground() whenever it feels like, there is only the single intent for startForegroundService (hard coded) that is passed to stopService.

I'll update the api again, but I really want to get this PR merged upstream because the reason I am using gio is to avoid having to muck about with android platform internals...

Reply:
> I don't understand why you think there are multiple foreground processes - GioForegroundService is a singleton

I didn't mean to imply multiple foreground services would exist. I meant that multiple background computations may happen concurrently. For example, let's say a component in your app is compressing a video while another is receiving new messages from a server. Those two processes are not related, but how would they coordinate their calls to Start and Stop?

I apologize for not being clearer.